deployer/src/deployment/run.rs

```diff
@@ -269,43 +277,191 @@ impl Built {
+
         // Execute loaded service
-        load(
+        let (old_resources, resources, new_secrets) = load(
             self.service_name.clone(),
             self.service_id,
-            executable_path.clone(),
-            resource_manager,
+            resource_manager.clone(),
             runtime_client.clone(),
-            self.claim,
+            self.claim.clone(),
             self.secrets,
         )
         .await?;
 
+        let resources = provision(
+            self.service_name.clone(),
+            self.service_id,
+            provisioner_client,
+            resource_manager,
+            self.claim,
+            old_resources,
+            resources,
+            new_secrets,
+        )
+        .await
+        .map_err(Error::Provision)?;
+
         let handler = tokio::spawn(run(
             self.id,
             self.service_name,
             runtime_client,
             address,
             cleanup,
+            resources,
         ));
 
         Ok(handler)
     }
 }
 
+async fn provision(
+    service_name: String,
+    service_id: Ulid,
+    mut provisioner_client: provisioner::Client,
+    mut resource_manager: impl ResourceManager,
+    claim: Claim,
+    old_resources: Vec<resource::Response>,
+    mut resources: Vec<Vec<u8>>,
+    new_secrets: HashMap<String, String>,
+) -> anyhow::Result<Vec<Vec<u8>>> {
+    let mut resources_to_save: Vec<record_request::Resource> = Vec::new();
+    for r in resources.iter_mut() {
+        match serde_json::from_slice::<ResourceInput>(r.as_slice())
+            .context("deserializing resource input")?
+        {
+            ResourceInput::Shuttle(shuttle_resource) => {
+                macro_rules! log {
+                    ($msg:expr) => {
+                        info!("[Resource][{}] {}", shuttle_resource.r#type, $msg);
+                    };
+                }
+                macro_rules! get_cached_output_or_provision {
+                    ($output_type:ty, $provision_logic:expr) => {{
+                        let output = old_resources
+                            .iter()
+                            .find(|resource| {
+                                resource.r#type == shuttle_resource.r#type
+                                    && resource.config == shuttle_resource.config
+                            })
+                            .and_then(|resource| {
+                                let cached_output = resource.data.clone();
+                                log!("Found cached output");
+                                match serde_json::from_value::<$output_type>(cached_output) {
+                                    Ok(output) => Some(output),
+                                    Err(_) => {
+                                        log!("Failed to validate cached output");
+                                        None
+                                    }
+                                }
+                            });
```

Comment:
I don't think this needs to be a macro. You should be able to refactor it into a function instead.

Reply:
my unfinished attempt: https://github.com/shuttle-hq/shuttle/pull/1628/commits/8c726055943354f5e070f790fe46e7b020eff436
feel free to try it out